nixos/modules/tasks/filesystems/zfs.nix

```
@@ -517,9 +522,18 @@ in
               done
               poolImported "${pool}" || poolImport "${pool}"  # Try one last time, e.g. to import a degraded pool.
               if poolImported "${pool}"; then
-                ${optionalString cfgZfs.requestEncryptionCredentials ''
+                ${optionalString (if isBool cfgZfs.requestEncryptionCredentials
+                                  then cfgZfs.requestEncryptionCredentials
+                                  else cfgZfs.requestEncryptionCredentials != []) ''
                   ${packages.zfsUser}/sbin/zfs list -rHo name,keylocation ${pool} | while IFS=$'\t' read ds kl; do
-                    (case "$kl" in
+                    (${if isBool cfgZfs.requestEncryptionCredentials
+                       then "" # requestEncryptionCredentials was true, so we loop over all (we insert no check)
+                       else ''
+                         if ! echo '${concatStringsSep cfgZfs.requestEncryptionCredentials "\n"}' | grep -qx "$ds"; then
+                           continue
+                         fi
+                       ''}
```

A bit more readable especially when you have to debug stage1 in a rescue shell.
```suggestion
                    (${optionalString (!isBool) cfgZfs.requestEncryptionCredentials ''
                      pools="${toString cfgZfs.requestEncryptionCredentials}"
                      if ! echo "$pools" | grep -qx "$ds"; then
                        continue
                      fi
                    ''}
```

---

Thank you for the suggestion, however, the `concatStringsSep` is still necessary.  Otherwise a list like `[ "rpool/data/home" "rpool/data/persist" ]` will be flattened to `"rpool/data/home rpool/data/persist"` by `toString` and then `grep` will never match because the `-x` option is equivalent to being anchored on the line, i.e. `^pattern$'`.  So one possibility is:
```suggestion
                    (${optionalString (!isBool) cfgZfs.requestEncryptionCredentials ''
                         if ! echo "${concatStringsSep cfgZfs.requestEncryptionCredentials "\n"}" | grep -qx "$ds"; then
                           continue
                         fi
                       ''}
```
The other possibility is to match surrounding whitespace in `grep`:
```suggestion
                    (${optionalString (!isBool) cfgZfs.requestEncryptionCredentials ''
                         if ! echo " ${toString cfgZfs.requestEncryptionCredentials} " | grep -q " $ds "; then
                           continue
                         fi
                       ''}
```

---

Your first suggestion looks more readable.